crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -413,12 +415,75 @@ impl<'db> TypeInferenceBuilder<'db> {
             for definition in self.types.declarations.keys() {
                 if infer_definition_types(self.db, *definition).has_deferred {
                     let deferred = infer_deferred_types(self.db, *definition);
-                    deferred_expression_types.extend(deferred.expressions.iter());
+                    deferred_expression_types.extend(&deferred.expressions);
                 }
             }
-            self.types
-                .expressions
-                .extend(deferred_expression_types.iter());
+            self.types.expressions.extend(deferred_expression_types);
+        }
+
+        self.check_class_definitions();
+    }
+
+    /// Iterate over all class definitions to check that Python will be able to create
+    /// a consistent "[method resolution order]" for each class at runtime. If not,
+    /// issue a diagnostic.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    fn check_class_definitions(&mut self) {
+        let class_definitions = self
+            .types
+            .declarations
+            .values()
+            .filter_map(|ty| ty.into_class_literal_type());
+
+        for class in class_definitions {
+            if class.is_cyclically_defined(self.db) {
+                self.diagnostics.add(
+                    class.node(self.db).into(),
+                    "cyclic-class-def",
+                    format_args!(
+                        "Cyclic definition of `{}` or bases of `{}` (class cannot inherit from itself)",
+                        class.name(self.db),
+                        class.name(self.db)
+                    )
+                );
+            }
+
+            match class.try_mro(self.db).as_ref().map_err(|err| &err.kind) {
```

Comment:
Considering that we're always calling `try_mro` here, could we instead move the `is_cyclically_defined` into `try_mro` so that it is only called once and that all MRO error handling is centralized?